p2p/discovery/broadcast.go

```diff
@@ -0,0 +1,181 @@
+	// sent from.
+	listenSock, err := bindBroadcast(net.IPv4bcast, defaultPort)
+	if err != nil {
+		return nil, err
+	}
+
+	bs := &broadcastService{
+		host:          peerhost,
+		listenSock:    listenSock,
+		broadcastPort: defaultPort,
+	}
+
+	addrs, err := getDialableListenAddrs(peerhost)
+	if err != nil {
+		bs.Close()
+		return nil, err
+	} else {
+		for _, a := range addrs {
+			if a.IP.IsGlobalUnicast() {
+				// binding to the same address and port as the TCP service
+				// embeds that address and port into broadcast messages
+				otherSock, err := bindBroadcast(a.IP, a.Port)
+				if err == nil {
+					bs.sendSocks = append(bs.sendSocks, otherSock)
+				}
+			}
+		}
+	}
+
+	// fudge the interval to dampen oscillations with other nodes
+	margin := int(interval) >> 2
+	interval += time.Duration(rand.Intn(margin))
+	interval -= time.Duration(rand.Intn(margin))
+
+	go bs.sendBroadcasts(ctx, interval)
+	go bs.recvBroadcasts(ctx)
+
+	return bs, nil
+}
+
+func (bs *broadcastService) Close() error {
+	bs.listenSock.Close()
+	for _, conn := range bs.sendSocks {
+		conn.Close()
+	}
+	return nil
+}
+
+func (bs *broadcastService) handleBroadcast(addr *net.UDPAddr, bpeer peer.ID) {
+	// Broadcast messages are UDP, but represent IPv4 TCP sockets
+	maddr, err := manet.FromNetAddr(&net.TCPAddr{
+		IP:   addr.IP,
+		Port: addr.Port,
+	})
+	if err == nil {
+		pi := pstore.PeerInfo{
+			ID:    bpeer,
+			Addrs: []ma.Multiaddr{maddr},
+		}
+
+		bs.lk.Lock()
+		for _, n := range bs.notifees {
+			go n.HandlePeerFound(pi)
+		}
+		bs.lk.Unlock()
+	}
+}
+
+func (bs *broadcastService) recvBroadcasts(ctx context.Context) {
+	var buf = make([]byte, 48)
+	for {
+		n, rAddr, err := bs.listenSock.ReadFromUDP(buf)
+		if err == nil {
+			if n != 0 {
+				rPeer, err := peer.IDB58Decode(string(buf[0:n]))
+				if err == nil && rPeer != bs.host.ID() {
+					bs.handleBroadcast(rAddr, rPeer)
+				}
+			}
+		}
+	}
+}
+
+func (bs *broadcastService) sendBroadcast() {
+	idBuf := []byte(bs.host.ID().Pretty())
+	broadcastAddr := net.UDPAddr{
+		IP:   net.IPv4bcast,
+		Port: bs.broadcastPort,
+	}
+	for _, conn := range bs.sendSocks {
+		conn.WriteToUDP(idBuf, &broadcastAddr)
+	}
+}
+
+func (bs *broadcastService) sendBroadcasts(ctx context.Context, interval time.Duration) {
+	// send an initial broadcast after a quarter interval
+	time.Sleep(interval >> 2)
+	bs.sendBroadcast()
+
+	// send broadcasts on a periodic timeout
+	ticker := time.NewTicker(interval)
+	for {
+		select {
+		case <-ticker.C:
+			bs.sendBroadcast()
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (m *broadcastService) RegisterNotifee(n Notifee) {
+	m.lk.Lock()
+	m.notifees = append(m.notifees, n)
+	m.lk.Unlock()
+}
+
+func (m *broadcastService) UnregisterNotifee(n Notifee) {
+	m.lk.Lock()
+	found := -1
+	for i, notif := range m.notifees {
+		if notif == n {
+			found = i
+			break
+		}
+	}
+	if found != -1 {
+		m.notifees = append(m.notifees[:found], m.notifees[found+1:]...)
```

Comment:
Since we don't really care about order, you can just swap the entry you want to remove for the last entry. Also, remember to null out the last entry, otherwise, it won't be garbage collected.